ECS - Run Task (Fargate)

grejdi-mbta opened the pull request.

grejdi-mbta committed:
New action for running ECS tasks on Fargate, and waiting for them to finish.

grejdi-mbta committed:
fix yamllint issues

paulswartz reviewed (commented):
🍰 looks fine to me; is there a workflow which uses this?

grejdi-mbta commented:
@paulswartz The data platform repo will use it to run the task for deploying its db migrations (see [this](https://github.com/mbta/data_platform/blob/main/.github/workflows/deploy.yml)). Still working out some more details.

grejdi-mbta committed:
make running task more flexible

grejdi-mbta commented:
@paulswartz small updates here to make things a little easier with less variables.

paulswartz commented on run-ecs-task/README.md:
```diff
@@ -0,0 +1,20 @@
+* **aws_default_region**: (Required) Will be passed to GitHub Actions environment as the AWS_DEFAULT_REGION environment variable to be used in AWS CLI commands (from `aws configure`).
+* **ecs_cluster**: (Required) Cluster ARN to be used for running the task (ex. `arn:aws:ecs:{aws-region}:{aws-account-id}:cluster/{cluster-name}`). 
+* **ecs_task_definition**: (Required) Task Definition ARN to base the task on (ex. `arn:aws:ecs:{aws-region}:{aws-account-id}:task-definition/{task-definition-name}`).
+* **ecs_task_network_configuration**: (Required) Network settings that the task will be run with. See [this](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-configure-network.html) for more information.
```
question: should this be required? it's required for tasks which use the `awsvpc` networking stack, but isn't supported for any other modes.

paulswartz commented on run-ecs-task/README.md:
```diff
@@ -0,0 +1,20 @@
+* **ecs_cluster**: (Required) Cluster ARN to be used for running the task (ex. `arn:aws:ecs:{aws-region}:{aws-account-id}:cluster/{cluster-name}`). 
+* **ecs_task_definition**: (Required) Task Definition ARN to base the task on (ex. `arn:aws:ecs:{aws-region}:{aws-account-id}:task-definition/{task-definition-name}`).
+* **ecs_task_network_configuration**: (Required) Network settings that the task will be run with. See [this](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-configure-network.html) for more information.
+* **ecs_task_overrides**: (Required) Override settings that the task will be run with. See [this](https://docs.aws.amazon.com/cli/latest/reference/ecs/run-task.html) for more information.
```
question: does this need to be required? If you wanted to keep the script simpler, you could default it to `[]`

thecristen commented on run-ecs-task/action.sh:
```diff
@@ -0,0 +1,45 @@
+# run task from task definition
+task_run=$(aws ecs run-task --task-definition "${ECS_TASK_DEFINITION}" --cluster "${ECS_CLUSTER}" --launch-type FARGATE --network-configuration "${ECS_TASK_NETWORK_CONFIGURATION}" --overrides "${ECS_TASK_OVERRIDES}")
+# get unique task arn
+task_arn=$(echo $task_run | jq '.tasks[0].taskArn')
```
```suggestion
task_arn=$(echo "$task_run" | jq '.tasks[0].taskArn')
```

thecristen commented on run-ecs-task/action.sh:
```diff
@@ -0,0 +1,45 @@
+while [[ "${finished}" == "false" && "${timed_out}" == "false" ]] ; do
+
+  # get the latest state of the task
+  task_description=$(aws ecs describe-tasks --cluster "${ECS_CLUSTER}" --tasks [$task_arn])
```
```suggestion
  task_description=$(aws ecs describe-tasks --cluster "${ECS_CLUSTER}" --tasks ["$task_arn"])
```

thecristen commented on run-ecs-task/action.sh:
```diff
@@ -0,0 +1,45 @@
+  # get the latest state of the task
+  task_description=$(aws ecs describe-tasks --cluster "${ECS_CLUSTER}" --tasks [$task_arn])
+  # get the latest status
+  task_status=$(echo $task_description | jq '.tasks[0].lastStatus')
```
```suggestion
  task_status=$(echo "$task_description" | jq '.tasks[0].lastStatus')
```

thecristen commented on run-ecs-task/action.sh:
```diff
@@ -0,0 +1,45 @@
+if [[ "${timed_out}" == "true" ]] ; then
+  echo 'Run Task failed!'
+  echo '----------------------------------------'
+  echo $task_description
```
```suggestion
  echo "$task_description"
```

thecristen reviewed (changes requested):
Some tiny formatting suggestions via [Shellcheck](https://www.shellcheck.net), and I echo Paul's questions about the required variables. Otherwise lgtm

grejdi-mbta committed:
Apply suggestions from code review

Co-authored-by: Cristen Jones <cjones3@mbta.com>

grejdi-mbta committed:
Apply suggestions from code review

Co-authored-by: Cristen Jones <cjones3@mbta.com>

grejdi-mbta committed:
remove required setting for network configuration and container overrides.

grejdi-mbta commented:
@paulswartz @thecristen Thanks both. I removed the 'required' as per your suggestion.

thecristen commented on run-ecs-task/action.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+# run task from task definition
+task_run=$(aws ecs run-task --task-definition "${ECS_TASK_DEFINITION}" --cluster "${ECS_CLUSTER}" --launch-type FARGATE --network-configuration "${ECS_TASK_NETWORK_CONFIGURATION}" --overrides "${ECS_TASK_OVERRIDES}")
```
Will this still work as expected if the inputs populating `ECS_TASK_NETWORK_CONFIGURATION` and `ECS_TASK_OVERRIDES` are undefined?

grejdi-mbta replied:
Will check. It was working for my use case, but it's a different setup from MBTA's infrastructure.

grejdi-mbta committed:
rename to be more specific. and add check for overrides.

grejdi-mbta commented:
@paulswartz @thecristen Getting back on this. I renamed this action to be more specific to just Fargate tasks. For Fargate tasks, the network configuration is **required**, but overrides is not. For overrides, I've added a check that runs the command with or without depending on the environment variable being available.

paulswartz reviewed (approved):
🍰 

You'll need to tag a new version, as well as re-tagging the `v1` tag.

grejdi-mbta commented:
@paulswartz Tagged, and updated `v1`. Thanks.
